File: src/features/analytics/__tests__/transforms.test.ts
import {
  computeExerciseRanks,
  frequencyByWeek,
  monthlyComparison,
  startOfWeekMs,
  volumeByWeek,
} from '@/features/analytics/transforms';
import type { SessionSummary } from '@/features/analytics/types';

// Fixed reference: Wed 2026-08-26 12:00 local. Week starts Mon 2026-08-24.
const NOW = new Date(2026, 7, 26, 12).getTime();

function session(daysAgo: number, volumeKg: number): SessionSummary {
  return {
    id: `s${daysAgo}`,
    name: 'W',
    completedAt: new Date(NOW - daysAgo * 86_400_000).toISOString(),
    gateClearRank: null,
    totalVolumeKg: volumeKg,
    durationSeconds: 0,
  };
}

describe('startOfWeekMs', () => {
  it('snaps to the Monday of the week', () => {
    const mon = new Date(startOfWeekMs(NOW));
    expect(mon.getDay()).toBe(1); // Monday
    expect(mon.getDate()).toBe(24);
  });
});

describe('volumeByWeek', () => {
  it('buckets volume per week and pads empty weeks', () => {
    const sessions = [session(0, 1000), session(8, 500)]; // this week, prev week
    const out = volumeByWeek(sessions, 3, NOW);
    expect(out).toHaveLength(3);
    expect(out.map((p) => p.value)).toEqual([0, 500, 1000]);
  });
  it('excludes sessions older than the window', () => {
    const out = volumeByWeek([session(60, 9999)], 3, NOW);
    expect(out.every((p) => p.value === 0)).toBe(true);
  });
});

describe('frequencyByWeek', () => {
  it('counts workouts per week', () => {
    const out = frequencyByWeek([session(0, 100), session(1, 100), session(8, 100)], 2, NOW);
    expect(out.map((p) => p.value)).toEqual([1, 2]);
  });
});

describe('monthlyComparison', () => {
  it('splits this month vs last month', () => {
    const sessions = [
      session(1, 1000), // Aug (this month)
      {
        ...session(0, 0),
        completedAt: new Date(2026, 6, 15, 12).toISOString(),
        totalVolumeKg: 2000,
      }, // Jul
      {
        ...session(0, 0),
        completedAt: new Date(2026, 5, 10, 12).toISOString(),
        totalVolumeKg: 5000,
      }, // Jun (neither)
    ];
    const c = monthlyComparison(sessions, NOW);
    expect(c.thisMonth).toEqual({ workouts: 1, volumeKg: 1000 });
    expect(c.lastMonth).toEqual({ workouts: 1, volumeKg: 2000 });
  });
});

describe('computeExerciseRanks', () => {
  it('ranks scorable lifts, drops unranked, sorts by score desc', () => {
    const ranks = computeExerciseRanks(
      [
        { exerciseId: 'e1', name: 'Barbell Bench Press', best1RMkg: 80 }, // ratio 1.0 -> 50 (B)
        { exerciseId: 'e2', name: 'Lateral Raise', best1RMkg: 30 }, // no standard -> dropped
        { exerciseId: 'e3', name: 'Barbell Back Squat', best1RMkg: 160 }, // ratio 2.0 -> ~86 (S)
      ],
      80,
      'male',
    );
    expect(ranks.map((r) => r.name)).toEqual(['Barbell Back Squat', 'Barbell Bench Press']);
    expect(ranks[1].rank).toBe('B');
    expect(ranks[0].rank).toBe('S');
  });
});

File: android-native/app/src/test/java/com/fitnessrpg/app/domain/analytics/TransformsTest.kt
package com.fitnessrpg.app.domain.analytics

import com.fitnessrpg.app.domain.rank.Rank
import com.fitnessrpg.app.util.isoFromMillis
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId

class TransformsTest {

    // Fixed reference: Wed 2026-08-26 12:00 local. Week starts Mon 2026-08-24.
    private val now: Long =
        LocalDateTime.of(2026, 8, 26, 12, 0).atZone(ZoneId.systemDefault()).toInstant().toEpochMilli()

    private fun session(daysAgo: Int, volumeKg: Double) = SessionSummary(
        id = "s$daysAgo",
        name = "W",
        completedAt = isoFromMillis(now - daysAgo * 86_400_000L),
        gateClearRank = null,
        totalVolumeKg = volumeKg,
        durationSeconds = 0,
    )

    private fun atNoon(y: Int, m: Int, d: Int): Long =
        LocalDateTime.of(y, m, d, 12, 0).atZone(ZoneId.systemDefault()).toInstant().toEpochMilli()

    @Test
    fun `startOfWeekMs snaps to the Monday of the week`() {
        val monday = Instant.ofEpochMilli(startOfWeekMs(now)).atZone(ZoneId.systemDefault()).toLocalDate()
        assertEquals(DayOfWeek.MONDAY, monday.dayOfWeek)
        assertEquals(24, monday.dayOfMonth)
    }

    @Test
    fun `volumeByWeek buckets volume per week and pads empty weeks`() {
        val sessions = listOf(session(0, 1000.0), session(8, 500.0))
        val out = volumeByWeek(sessions, 3, now)
        assertEquals(3, out.size)
        assertEquals(listOf(0, 500, 1000), out.map { it.value })
    }

    @Test
    fun `volumeByWeek excludes sessions older than the window`() {
        val out = volumeByWeek(listOf(session(60, 9999.0)), 3, now)
        assertTrue(out.all { it.value == 0 })
    }

    @Test
    fun `frequencyByWeek counts workouts per week`() {
        val out = frequencyByWeek(listOf(session(0, 100.0), session(1, 100.0), session(8, 100.0)), 2, now)
        assertEquals(listOf(1, 2), out.map { it.value })
    }

    @Test
    fun `monthlyComparison splits this month vs last month`() {
        val sessions = listOf(
            session(1, 1000.0), // Aug (this month)
            SessionSummary("jul", "W", isoFromMillis(atNoon(2026, 7, 15)), null, 2000.0, 0),
            SessionSummary("jun", "W", isoFromMillis(atNoon(2026, 6, 10)), null, 5000.0, 0),
        )
        val c = monthlyComparison(sessions, now)
        assertEquals(MonthTotals(1, 1000), c.thisMonth)
        assertEquals(MonthTotals(1, 2000), c.lastMonth)
    }

    @Test
    fun `computeExerciseRanks ranks scorable lifts, drops unranked, sorts by score desc`() {
        val ranks = computeExerciseRanks(
            listOf(
                ExerciseStatInput("e1", "Barbell Bench Press", 80.0), // ratio 1.0 -> 50 (B)
                ExerciseStatInput("e2", "Lateral Raise", 30.0), // no standard -> dropped
                ExerciseStatInput("e3", "Barbell Back Squat", 160.0), // ratio 2.0 -> ~86 (S)
            ),
            80.0,
            "male",
        )
        assertEquals(listOf("Barbell Back Squat", "Barbell Bench Press"), ranks.map { it.name })
        assertEquals(Rank.B, ranks[1].rank)
        assertEquals(Rank.S, ranks[0].rank)
    }
}
